### PR TITLE
fix(agent-sessions): scope Codex subagent faults

### PR DESCRIPTION
**Pull request by Maxsky5**
## Summary

- derive Codex subagent lineage only from authoritative spawn, fork, and collaboration events
- attach exact parent or child ownership before fallible event projection and route child diagnostics to the owning session
- propagate projection callback failures to the host boundary and emit structured scoped or unscoped fault logs independently from live-envelope publication
- track transient frontend faults per session so switching sessions does not leak an unrelated failure, and clear recovered transcript faults
- split request-only and streaming Codex adapter options so only live stream registrations require queue-failure handling

## Motivation and context

Codex subagent activity could previously infer invalid child links from non-authoritative events. When projection failed, the resulting error could lack the exact session owner, contaminate whichever session was selected in Agent Studio, or be disconnected from the host logging path. This made a child-session failure appear as a broader session or UI failure and made recovery behavior unreliable.

The change keeps runtime lineage and fault ownership explicit from the adapter through the host live-state service to the frontend read model. It does not add persisted fields or alter the SQLite task/session schema.

## Technical choices

- treat spawn/fork/collaboration payloads as authoritative lineage sources rather than normalizing arbitrary event references into links
- resolve routed child ownership before event preprocessing so mapping failures retain the correct fault reference
- surface asynchronous Codex projection failures through an explicit streaming callback instead of swallowing or retrying them
- fan out host logging and live-envelope publication independently so a logging failure cannot suppress fault publication
- use canonical session identity for transient frontend fault state and clear faults when transcript activity demonstrates recovery
- preserve fail-fast behavior; no fallback routes, polling, retries, or durable runtime metadata were introduced

## Verification

Run after rebasing onto current `origin/main`:

- `bun run lint`
- `bun run typecheck`
- `bun run test`
- `bun run build`
- `git diff --check origin/main...HEAD`

All passed. The build retains existing chunk-size warnings, and the frontend test suite retains existing non-failing React `act(...)` diagnostics. Cargo checks are not applicable because this repository tracks no Rust manifests or source files.

## Summary by CodeRabbit

* **New Features**
  * Added per-session auxiliary error reporting in Agent Studio, including transient fault support and improved Retry suppression when auxiliary errors exist.
  * Improved routing of runtime stream events and pending request lifecycle events to the correct target session, supporting nested/forked flows.
  * Enhanced descendant pending-input projection to mirror items across active ancestor chains.

* **Bug Fixes**
  * Fixed mis-attribution of faults, approvals, and questions to the wrong session.
  * Improved routed snapshot retention for deeper descendant trees with deduping.
  * Strengthened fault logging/publishing and scoping, including malformed and cross-runtime cases.

* **Tests**
  * Expanded coverage for routing, scoping, and fault-handling edge cases.

### CODE REVIEW

File: packages/host/src/application/agent-sessions/agent-session-live-state-service.ts [hunk 1]
```diff
@@ -136,27 +139,90 @@ const toEnvelope = (change: AgentSessionLiveAdapterChange): AgentSessionLiveEnve
...
+                faultLogFailure: faultLogResult.left,
+                publishFailure: publishResult.left,
+              },
             }),
+          );
+        }
+        if (faultLogResult._tag === "Left") {
+          return yield* Effect.fail(faultLogResult.left);
+        }
+        if (publishResult._tag === "Left") {
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply packages/host/src/application/agent-sessions/agent-session-live-state-service.ts:217]: Fixed in `06cef20`. Logging-only failures are retained while later envelopes continue publishing, then returned after the batch. Validation or publication failures remain terminal so no deltas publish after a stream gap. Added `[fault, session_upsert]` continuation and publication-gap stop coverage; full checks pass.

File: packages/adapters-codex-app-server/src/codex-runtime-session-events.ts [hunk 2]
```diff
@@ -201,12 +228,15 @@ export class CodexRuntimeSessionEvents {
...
+  private enqueueRuntimeStreamEvent(
+    event: CodexRuntimeStreamEvent,
+    onRuntimeEventQueueFailure: CodexRuntimeEventQueueFailureHandler,
+  ): void {
     const previous =
       this.runtimeEventProcessingByRuntimeId.get(event.runtimeId) ?? Promise.resolve();
     const processing = previous
       .then(() => this.processRuntimeStreamEventMutation(event))
-      .catch((error) => this.emitRuntimeStreamEventError(event, error));
+      .catch((error) => onRuntimeEventQueueFailure({ runtimeId: event.runtimeId, error }));
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply packages/adapters-codex-app-server/src/codex-runtime-session-events.ts]: Fixed in `06cef20`. Queue failure reporting is now a synchronous terminal contract owned by the Effect runtime, while the per-runtime map stores an always-settled queue tail. The original processing failure is reported once and an already-queued same-runtime successor still runs. Added deferred queue and dual-delivery regression coverage; full lint, typecheck, test, and build checks pass.

File: packages/adapters-codex-app-server/src/codex-runtime-session-events.ts [hunk 3]
```diff
@@ -542,17 +576,31 @@ export class CodexRuntimeSessionEvents {
...
-      return;
+      return { faultSession: session };
     }
     const route = this.deps.subagents.routeForChild(threadId, event.runtimeId);
     const parentSession = route ? this.deps.sessions.get(route.parentExternalSessionId) : undefined;
-    if (parentSession?.runtimeId === event.runtimeId) {
-      this.emitSessionError(parentSession.threadId, error);
-      return;
+    if (!route || parentSession?.runtimeId !== event.runtimeId) {
+      return undefined;
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply packages/adapters-codex-app-server/src/codex-runtime-session-events.ts]: Fixed in `06cef20`. Runtime ownership now walks same-runtime ancestry to a retained ancestor while keeping the original nested target identity. Routed descendants are projected recursively, and pending/resolved approval and question events remain observable without retaining synthetic sessions. Added nested, cross-runtime, direct-session, and retained parent/child delivery coverage; full checks pass.

**Comment by Maxsky5** [comment]
Addressed the actionable FriendlyReviewer, Greptile, and Codex feedback in `06cef20`.

**Fixed**
- queue reporter failures can no longer poison later runtime events
- catalog invalidation and live mutation delivery are both attempted once, with actionable aggregate failure reporting
- nested child faults and request lifecycle events retain the original child identity
- fault-log-only failures no longer truncate later host batch publication
- request-only/streaming test fixtures now respect the discriminated contract
- simultaneous exact-session and read-model failures remain visible, with the exact session fault still primary

**Intentionally unchanged**
- destructured union narrowing: current TypeScript narrows it correctly and full typecheck passes
- the unscoped `routeForChild` assertion: that fixture intentionally records an unscoped route; runtime-scoped routing has separate coverage
- unroutable `server_request` broadcast: pre-existing policy that needs a separate global diagnostics contract; removing it here could hide failures
- synthetic child storage: routed children remain live projections by design; ancestry and recursive snapshots now provide ownership without duplicating lifecycle state
- the lower subscription guard: retained as boundary validation

CodeRabbit produced only a rate-limit status and React Doctor reported no new issues, so those required no code change. Verification passed: React Doctor 100/100, `bun run lint`, `bun run typecheck`, `bun run test`, and `bun run build`.

File: packages/adapters-codex-app-server/src/codex-app-server-adapter.ts [hunk 4]
```diff
@@ -548,13 +554,22 @@ export class CodexAppServerAdapter
...
+      parentExternalSessionId: string,
+    ): void => {
+      for (const route of this.subagents.routesForParent(parentExternalSessionId, runtimeId)) {
+        if (visited.has(route.childExternalSessionId)) {
           continue;
         }
-        snapshots.push(this.toRoutedChildLiveSessionSnapshot(parentSession, route));
+        visited.add(route.childExternalSessionId);
+        snapshots.push(this.toRoutedChildLiveSessionSnapshot(retainedAncestor, route));
+        appendRoutedDescendants(retainedAncestor, route.childExternalSessionId);
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply packages/adapters-codex-app-server/src/codex-app-server-adapter.ts:568]: Fixed in `ac6deaf`. The lifecycle projector now updates the routed child link before returning when its immediate parent is unretained, so recursively projected grandchildren no longer remain `running`. Parent transcript emission stays buffered and replays once the parent becomes retained. Added a focused `root -> child -> grandchild` regression; full lint, typecheck, test, and build checks pass.

File: packages/adapters-codex-app-server/src/codex-subagent-lifecycle-projector.ts [hunk 5]
```diff
@@ -155,6 +152,9 @@ export class CodexSubagentLifecycleProjector {
         ? { startedAtMs: update.timestampMs }
         : { endedAtMs: update.timestampMs }),
     });
+    if (!parentSession) {
+      return false;
+    }
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply packages/adapters-codex-app-server/src/codex-subagent-lifecycle-projector.ts:157]: Verified against the current production path. Buffered replay is initiated from `thread/started`, whose projected lifecycle is terminal (`completed`); there is no production caller replaying a buffered non-terminal `running` lifecycle. The unchanged-status branch intentionally deduplicates already-emitted live updates. Forcing an emission or adding emitted-state tracking would add complexity for an unreachable path, so this is intentionally deferred until a real non-terminal replay path exists.

File: packages/adapters-codex-app-server/src/codex-app-server-adapter.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (Maxsky5) [reply packages/adapters-codex-app-server/src/codex-app-server-adapter.ts:568]: Fixed in `874ad7b79`. `CodexContextUsageLoader` now walks the runtime-scoped parent route chain to the first retained ancestor with cycle and cross-runtime guards, while resuming/loading context for the original grandchild thread. Added retained-root grandchild success and invalid-chain no-resume coverage. Focused tests: 7/7; full monorepo tests and build pass.

[the remark this answers is not in the document]
Comment (Maxsky5) [reply packages/adapters-codex-app-server/src/codex-app-server-adapter.ts:568]: Fixed in `874ad7b79`. The frontend now keeps a transient live parent link and rebuilds each direct pending approval/question onto every live ancestor while preserving the original owner as `responseSession`. Added coverage for read-only policy rejection, questions, resolution/removal clearing, sibling isolation, and one-hop compatibility. The field is excluded from persistence. Focused tests: 13/13; full monorepo tests and build pass.
